src/main/java/org/apache/commons/lang3/StringUtils.java

```diff
@@ -7968,9 +7968,145 @@ public static String rotate(final String str, final int shift) {
             }
         }
         if (match || preserveAllTokens && lastMatch) {
-            list.add(str.substring(start, i));
+            buffer.add(str.substring(start, i));
+        }
+        return buffer.toArray();
+    }
+
+    private static final ThreadLocal<SplitBufferThreadLocalHelper> SPLIT_BUFFER_THREAD_LOCAL
+            = new ThreadLocal<SplitBufferThreadLocalHelper>() {
+        @Override
+        protected SplitBufferThreadLocalHelper initialValue() {
+            return new SplitBufferThreadLocalHelper();
+        }
+    };
+
+    // Private class act as a buffer while splitting.
+    // "SplitBufferThreadLocalHelper" is constructed as a thread local variable so it is
+    // thread safe. The "splitBuffer" field acts as a buffer to hold the temporary
+    // representation of string split segments. It is shared by all
+    // calls to splitByCharacterType() or splitByWholeSeparatorWorker()
+    // or splitWorker(String) or splitWorker(char) and its variants in that particular thread.
+    private static final class SplitBufferThreadLocalHelper {
+
+        private final SplitBuffer splitBuffer = new SplitBuffer();
+
+        SplitBuffer getBuffer(){
+            splitBuffer.reset();
+            return splitBuffer;
+        }
+    }
+
+    //buffer class
+    private static final class SplitBuffer {
```

Comment:
`StringUtils` is a long class, with high maintenance cost already as-is. Adding an extra class increases the maintenance. I would prefer to avoid it if possible. If we really need to add it, we **must** provide good documentation. Reading the name of the class, and the single comment `buffer class` gives too much room for interpretation. IOW, I would have to read the code to understand what this class does, how it works, and whether I should use it or not.